src/pymoca/tree.py

```diff
@@ -259,6 +290,983 @@ def handle_walk(self, listener: TreeListener, tree: Union[ast.Node, dict, list])
+    # among the declared named elements of the temporary flattened class. If
+    # the class does not satisfy the requirements for a package, the lookup
+    # is restricted to encapsulated elements only. The class we look inside
+    # shall not be partial in a simulation model."
+    # Why do we have to temporarily flatten the class? Flattening requires name lookup
+    # and with this name lookup requires flattening. Yikes! Can it be simplified?
+
+    # TODO: Per spec v3.5 section 5.3.2 bullet 4, class is temporarily flattened
+    # For now, we use recursive name lookup in contained elements
+    found = _find_name(rest_of_name, first, search_parent=False)
+
+    # Check that found meets non-package lookup requirements in spec section 5.3.2
+    # The found.name test is so we only check going left to right in composite name
+    # and not the other direction as we pop the recursive call stack.
+    if (
+        found is not None
+        and found.name == _first_name(rest_of_name)
+        and first.type != "package"
+        and not (isinstance(found, ast.Class) and found.encapsulated)
+    ):
+        raise NameLookupError(f"{first.name} is not a package so {found.name} must be encapsulated")
+
+    return found
+
+
+def _first_name(name: str) -> str:
+    return name.split(".")[0]
+
+
+def _find_local(
+    name: str,
+    scope: ast.Class,
+) -> Optional[Union[ast.Class, ast.Symbol]]:
+    """Name lookup for predefined classes and contained elements"""
+
+    # 1. Iteration variables
+    # 2. Classes
+    # 3. Components (Symbols in Pymoca)
+
+    # 1. Iteration variables
+    # TODO: Refactor when handling iteration variables (it will move up one level)
+    if found := _find_iteration_variable(name, scope):
+        return found
+
+    # 2. Classes
+    if name in scope.classes:
+        return scope.classes[name]
+
+    # 3. Components (Symbols in Pymoca)
+    if name in scope.symbols:
+        return scope.symbols[name]
+
+    return None
+
+
+def _find_iteration_variable(name: str, scope: ast.Class) -> Optional[ast.Symbol]:
+    """Currently a pass"""
+    # TODO: Implement find name in iteration variables
+    return None
+
+
+def _find_inherited(
+    name: str,
+    scope: ast.Class,
+    current_extends: Optional[List[Union[ast.ExtendsClause, ast.InstanceExtends]]] = None,
+) -> Optional[Union[ast.Class, ast.Symbol]]:
+    """Find simple name in inherited classes"""
+    for extends in scope.extends:
+        if current_extends:
+            if extends in current_extends:
+                # We are in the middle of processing this one, don't do it infinitely :-)
+                continue
+        else:
+            current_extends = []
+        current_extends.append(extends)
+
+        if isinstance(extends, ast.InstanceExtends):
+            return _find_name(
+                name,
+                extends,
+                current_extends=current_extends,
+            )
+
+        extends_scope = _find_name(
```

Comment:
In `InstantiationTree`, what happens is that the name of the extends clause `extends TreeParts.Trunk` is looked up in the inherited class itself. That's a bit wasteful (and also confusing). Because we are essentially doing the same lookup twice.

Maybe we can add a check that we early terminate when `_find_inherited` is called for a `name` that's in `[x.component_ref for x in current_extends]` or something?